File: 2023/02/java/Main.java
package kival.aoc;

import java.util.ArrayList;
import java.util.List;

import kival.aoc.core.*;

public class Main {
   public final boolean HAS_IO = false;
public final boolean HAS_ALTERNATE = false;

   private class PogCube {
      public String type;
      public int amount;
   }

   private PogCube[] getCubes(String game) {
      String[] str = game.substring(game.indexOf(':') + 1).split(",|;");
      PogCube[] cubes = new PogCube[str.length];
      int i = 0;
      for (String chunk : str) {
         String[] val = chunk.trim().split(" ");
         PogCube cube = new PogCube();
         cube.type = val[1];
         cube.amount = Integer.parseInt(val[0]);
         cubes[i++] = cube;
      }
      return cubes;
   }

   public String part1(String input, boolean isTest) {
      int res = 0;
      int idx = 0;
      main: for (String line : input.split("\n")) {
         idx++;
         PogCube[] cubes = getCubes(line);
         for (PogCube cube : cubes) {
            if (cube.type.compareTo("red") == 0 && cube.amount > 12)
               continue main;
            if (cube.type.compareTo("green") == 0 && cube.amount > 13)
               continue main;
            if (cube.type.compareTo("blue") == 0 && cube.amount > 14)
               continue main;
         }
         res += idx;
      }
      return String.valueOf(res);
   }

   public String part2(String input, boolean isTest) {
      int res = 0;
      for (String line : input.split("\n")) {
         PogCube[] cubes = getCubes(line);
         int red = 0;
         int green = 0;
         int blue = 0;
         for (PogCube cube : cubes) {
            if (cube.type.compareTo("red") == 0 && cube.amount > red)
               red = cube.amount;
            if (cube.type.compareTo("green") == 0 && cube.amount > green)
               green = cube.amount;
            if (cube.type.compareTo("blue") == 0 && cube.amount > blue)
               blue = cube.amount;
         }
         res += red * green * blue;
      }
      return String.valueOf(res);
   }

   public void main(String[] args) {
      Run.execute(args, this::part1, this::part2, HAS_ALTERNATE, HAS_IO);
   }
}

File: 2023/02/go/main.go
package main

import (
	core "advent/of/go/src/go"
	"os"
	"strconv"
	"strings"
)

func main() {
	options := core.SolutionOptions{
		HasAlternate: false,
		HasIo:        false,
	}
	core.Run(os.Args, part1, part2, options)
}

type xdd struct {
	value int64
	color int64
}

func parseInput(input string) [][]xdd {
	lines := strings.Split(input, "\n")
	var results [][]xdd = make([][]xdd, 0)
	for i := 0; i < len(lines); i++ {
		line := strings.Split(strings.ReplaceAll(strings.Split(lines[i], ":")[1], ",", ";"), ";")
		cubes := make([]xdd, 0)
		for j := 0; j < len(line); j++ {
			lmao := strings.Split(strings.Trim(line[j], " "), " ")
			val, _ := strconv.ParseInt(lmao[0], 10, 64)
			col := int64(lmao[1][0]) % 3
			cubes = append(cubes, xdd{
				value: val,
				color: col,
			})
		}
		results = append(results, cubes)
	}
	return results
}

var rgb_p1 = [3]int64{12, 13, 14}

func part1(input string, isTest bool) string {
	parsed := parseInput(input)
	res := 0

	for i := 0; i < len(parsed); i++ {
		cubes := parsed[i]
		found := false
		for j := 0; j < len(cubes); j++ {
			cube := cubes[j]
			if cube.value > rgb_p1[cube.color] {
				found = true
			}
		}
		if !found {
			res += i + 1
		}
	}

	return strconv.Itoa(res)
}

func part2(input string, isTest bool) string {
	parsed := parseInput(input)
	var res int64 = 0

	for i := 0; i < len(parsed); i++ {
		cubes := parsed[i]
		rgb := [3]int64{0, 0, 0}
		for j := 0; j < len(cubes); j++ {
			cube := cubes[j]
			if cube.value > rgb[cube.color] {
				rgb[cube.color] = cube.value
			}
		}
		res += rgb[0] * rgb[1] * rgb[2]
	}

	return strconv.FormatInt(res, 10)
}
